File: frontend/src/types/todo.ts
/**
 * Todo(개인 할 일) 타입 정의
 */

export type TodoPriority = 'URGENT' | 'HIGH' | 'NORMAL' | 'LOW'
export type RepeatType = 'NONE' | 'DAILY' | 'WEEKLY' | 'MONTHLY' | 'YEARLY'
export type TodoSharePermission = 'VIEW' | 'EDIT'

/**
 * Todo 인터페이스
 */
export interface Todo {
  todoId: number
  ownerUserId: number
  ownerUserName?: string
  ownerUsername?: string
  title: string
  description?: string
  priority: TodoPriority
  dueDate?: string
  dueTime?: string
  isCompleted: boolean
  completedAt?: string
  sortOrder: number
  repeatType?: RepeatType
  repeatInterval?: number
  repeatDays?: string
  repeatEndDate?: string
  nextDueDate?: string
  parentTodoId?: number
  transferredFromUserId?: number
  transferredFromUserName?: string
  transferredAt?: string
  createdAt: string
  createdBy: string
  updatedAt?: string
  updatedBy?: string
  shareCount?: number
  isOverdue?: boolean
  isDueToday?: boolean
  shares?: TodoShare[]
}

/**
 * Todo 공유 인터페이스
 */
export interface TodoShare {
  shareId: number
  todoId: number
  todoTitle?: string
  sharedUserId: number
  sharedUserName?: string
  sharedUsername?: string
  departmentName?: string
  sharedByUserId: number
  sharedByUserName?: string
  sharedByUsername?: string
  permission: TodoSharePermission
  createdAt: string
  createdBy: string
  // Todo 관련 필드
  todoDueDate?: string
  todoDueTime?: string
  todoPriority?: TodoPriority
  todoOwnerUserName?: string
}

/**
 * Todo 생성 요청
 */
export interface TodoCreateRequest {
  title: string
  description?: string
  priority?: TodoPriority
  dueDate?: string
  dueTime?: string
  sortOrder?: number
  repeatType?: RepeatType
  repeatInterval?: number
  repeatDays?: string
  repeatEndDate?: string
}

/**
 * Todo 수정 요청
 */
export interface TodoUpdateRequest {
  title?: string
  description?: string
  priority?: TodoPriority
  dueDate?: string
  dueTime?: string
  isCompleted?: boolean
  sortOrder?: number
  repeatType?: RepeatType
  repeatInterval?: number
  repeatDays?: string
  repeatEndDate?: string
}

/**
 * Todo 공유 요청
 */
export interface TodoShareRequest {
  sharedUsername: string
  permission?: TodoSharePermission
}

/**
 * Todo 공유 권한 수정 요청
 */
export interface TodoShareUpdateRequest {
  permission: TodoSharePermission
}

/**
 * Todo 이관 요청
 */
export interface TodoTransferRequest {
  toUsername: string
  todoIds?: number[]
}

/**
 * Todo 일괄 이관 요청 (사용자 삭제용)
 */
export interface TodoBulkTransferRequest {
  fromUserId: number
  toUserId: number
  todoIds?: number[]
}

/**
 * Todo 순서 변경 요청
 */
export interface TodoReorderRequest {
  todoIds: number[]
}

/**
 * Todo 수 응답
 */
export interface TodoCountResponse {
  total: number
  active: number
}

/**
 * 우선순위 옵션
 */
export const PRIORITY_OPTIONS: { value: TodoPriority; label: string; color: string }[] = [
  { value: 'URGENT', label: '긴급', color: 'red' },
  { value: 'HIGH', label: '높음', color: 'orange' },
  { value: 'NORMAL', label: '보통', color: 'blue' },
  { value: 'LOW', label: '낮음', color: 'gray' }
]

/**
 * 반복 타입 옵션
 */
export const REPEAT_TYPE_OPTIONS: { value: RepeatType; label: string }[] = [
  { value: 'NONE', label: '반복 없음' },
  { value: 'DAILY', label: '매일' },
  { value: 'WEEKLY', label: '매주' },
  { value: 'MONTHLY', label: '매월' },
  { value: 'YEARLY', label: '매년' }
]

/**
 * 요일 옵션 (WEEKLY 반복용)
 */
export const WEEKDAY_OPTIONS: { value: string; label: string }[] = [
  { value: '1', label: '월' },
  { value: '2', label: '화' },
  { value: '3', label: '수' },
  { value: '4', label: '목' },
  { value: '5', label: '금' },
  { value: '6', label: '토' },
  { value: '7', label: '일' }
]

/**
 * 우선순위 레벨 반환
 */
export function getPriorityLevel(priority: TodoPriority): number {
  const levels: Record<TodoPriority, number> = {
    URGENT: 4,
    HIGH: 3,
    NORMAL: 2,
    LOW: 1
  }
  return levels[priority] || 0
}

/**
 * 우선순위 색상 반환
 */
export function getPriorityColor(priority: TodoPriority): string {
  const colors: Record<TodoPriority, string> = {
    URGENT: 'bg-red-500',
    HIGH: 'bg-orange-500',
    NORMAL: 'bg-blue-500',
    LOW: 'bg-gray-400'
  }
  return colors[priority] || 'bg-gray-400'
}

File: backend/src/main/java/com/taskflow/domain/Todo.java
package com.taskflow.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

/**
 * 개인 Todo 엔티티
 *
 * 테이블: TB_TODO
 */
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Todo {

    /**
     * Todo ID (PK)
     */
    private Long todoId;

    /**
     * 소유자 USER_ID (FK → TB_USER)
     */
    private Long ownerUserId;

    /**
     * Todo 제목
     */
    private String title;

    /**
     * 메모/상세 설명
     */
    private String description;

    /**
     * 우선순위 (URGENT/HIGH/NORMAL/LOW)
     */
    private String priority;

    /**
     * 마감일
     */
    private LocalDate dueDate;

    /**
     * 마감 시간
     */
    private LocalTime dueTime;

    /**
     * 완료 여부
     */
    private Boolean isCompleted;

    /**
     * 완료 일시
     */
    private LocalDateTime completedAt;

    /**
     * 정렬 순서
     */
    private Integer sortOrder;

    /**
     * 반복 유형 (NONE/DAILY/WEEKLY/MONTHLY/YEARLY)
     */
    private String repeatType;

    /**
     * 반복 간격
     */
    private Integer repeatInterval;

    /**
     * 반복 요일 (1,2,3,4,5)
     */
    private String repeatDays;

    /**
     * 반복 종료일
     */
    private LocalDate repeatEndDate;

    /**
     * 다음 반복 마감일
     */
    private LocalDate nextDueDate;

    /**
     * 반복 원본 TODO_ID (FK → TB_TODO)
     */
    private Long parentTodoId;

    /**
     * 이관 출처 사용자 (FK → TB_USER)
     */
    private Long transferredFromUserId;

    /**
     * 이관 일시
     */
    private LocalDateTime transferredAt;

    /**
     * 사용 여부
     */
    private String useYn;

    /**
     * 생성일시
     */
    private LocalDateTime createdAt;

    /**
     * 생성자 USERNAME
     */
    private String createdBy;

    /**
     * 수정일시
     */
    private LocalDateTime updatedAt;

    /**
     * 수정자 USERNAME
     */
    private String updatedBy;

    // =============================================
    // 추가 필드 (Mapper에서 JOIN으로 설정)
    // =============================================

    /**
     * 소유자명
     */
    private String ownerUserName;

    /**
     * 소유자 로그인ID
     */
    private String ownerUsername;

    /**
     * 이관 출처 사용자명
     */
    private String transferredFromUserName;

    /**
     * 공유 수
     */
    private Integer shareCount;

    // =============================================
    // 상수
    // =============================================

    public static final String PRIORITY_URGENT = "URGENT";
    public static final String PRIORITY_HIGH = "HIGH";
    public static final String PRIORITY_NORMAL = "NORMAL";
    public static final String PRIORITY_LOW = "LOW";

    public static final String REPEAT_NONE = "NONE";
    public static final String REPEAT_DAILY = "DAILY";
    public static final String REPEAT_WEEKLY = "WEEKLY";
    public static final String REPEAT_MONTHLY = "MONTHLY";
    public static final String REPEAT_YEARLY = "YEARLY";

    // =============================================
    // 편의 메서드
    // =============================================

    /**
     * 완료 여부 getter (Lombok 호환성)
     */
    public Boolean getIsCompleted() {
        return isCompleted;
    }

    /**
     * 반복 Todo 여부
     */
    public boolean hasRepeat() {
        return repeatType != null && !REPEAT_NONE.equals(repeatType);
    }

    /**
     * 이관받은 Todo 여부
     */
    public boolean isTransferred() {
        return transferredFromUserId != null;
    }

    /**
     * 반복 생성된 Todo 여부
     */
    public boolean isRepeatedInstance() {
        return parentTodoId != null;
    }

    /**
     * 우선순위 레벨 (정렬용)
     */
    public int getPriorityLevel() {
        if (PRIORITY_URGENT.equals(priority)) return 4;
        if (PRIORITY_HIGH.equals(priority)) return 3;
        if (PRIORITY_NORMAL.equals(priority)) return 2;
        if (PRIORITY_LOW.equals(priority)) return 1;
        return 0;
    }

    /**
     * 지연 여부
     */
    public boolean isOverdue() {
        if (Boolean.TRUE.equals(isCompleted)) return false;
        if (dueDate == null) return false;

        LocalDate today = LocalDate.now();
        if (dueDate.isBefore(today)) return true;

        if (dueDate.isEqual(today) && dueTime != null) {
            return dueTime.isBefore(LocalTime.now());
        }

        return false;
    }

    /**
     * 오늘 마감 여부
     */
    public boolean isDueToday() {
        if (dueDate == null) return false;
        return dueDate.isEqual(LocalDate.now());
    }

    // Lombok 호환성 getters
    public Long getTodoId() {
        return todoId;
    }

    public Long getOwnerUserId() {
        return ownerUserId;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public String getPriority() {
        return priority;
    }

    public LocalDate getDueDate() {
        return dueDate;
    }

    public LocalTime getDueTime() {
        return dueTime;
    }

    public LocalDateTime getCompletedAt() {
        return completedAt;
    }

    public Integer getSortOrder() {
        return sortOrder;
    }

    public String getRepeatType() {
        return repeatType;
    }

    public Integer getRepeatInterval() {
        return repeatInterval;
    }

    public String getRepeatDays() {
        return repeatDays;
    }

    public LocalDate getRepeatEndDate() {
        return repeatEndDate;
    }

    public LocalDate getNextDueDate() {
        return nextDueDate;
    }

    public Long getParentTodoId() {
        return parentTodoId;
    }

    public Long getTransferredFromUserId() {
        return transferredFromUserId;
    }

    public LocalDateTime getTransferredAt() {
        return transferredAt;
    }

    public String getUseYn() {
        return useYn;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public String getCreatedBy() {
        return createdBy;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public String getUpdatedBy() {
        return updatedBy;
    }

    public String getOwnerUserName() {
        return ownerUserName;
    }

    public String getOwnerUsername() {
        return ownerUsername;
    }

    public String getTransferredFromUserName() {
        return transferredFromUserName;
    }

    public Integer getShareCount() {
        return shareCount;
    }
}
